base/hashing.jl

```diff
@@ -24,6 +24,8 @@ hash(w::WeakRef, h::UInt) = hash(w.value, h)
 
 hash(@nospecialize(x), h::UInt) = hash_uint(3h - objectid(x))
 
+hash(x::Symbol) = objectid(x)
```

Comment:
We've never added these before, except for one case where someone overloaded the wrong method:
```
julia> methods(hash, (Any,))
# 2 methods for generic function "hash":
[1] hash(x::Base.Prehashed) in Base at multidimensional.jl:1553
[2] hash(x) in Base at hashing.jl:20
```

Reply:
Yes, usually you're supposed to overload the 2-arg one, but `Dict` calls `hash(x)`, and if you have a hash value pre-computed there's no reason to further call `hash_uint` on it. It's just wasted cycles. I think types with pre-computed hash values are a special case here, and could perhaps be mentioned in the docs even.

Reply:
Yes, no particular reason this current arrangement makes sense, but just observing that it seems to have always been considered incorrect so far. I'm not sure why we didn't instead design this to use one-arg hash, but I assumed there was a reason we don't change this to:
```julia
hash(@nospecialize(x)) = objectid(x)
hash(@nospecialize(x), h::UInt) = hash_uint(3h - hash(x))
```

Reply:
Yeah in hindsight that may have been better; the only advantage of the current setup is it makes life easier for types whose hash algorithm can take an input hash value.

Reply:
Before my big hashing revamp we did do single-argument hash. But then in order to hash collections, we had to hash each value and then call `bitmix` to combine the bits of the two hashes. This was inefficient since `bitmix` was itself a two-argument hashing function. The more efficient way to hash chains of values is to use a [Merkle–Damgård construction](https://en.wikipedia.org/wiki/Merkle–Damgård_construction), which requires a two-argument hashing function.

The alternative would be to make the `bitmix` function so cheap that it doesn't matter that it has to be called in addition to the single argument `hash` function. For that design it would be possible to use:
```jl
bitmix(a::UInt, b::UInt) = 3a - b
```
Where `a` would be the hash of already seen values and `b` would be the hash of the new value to be combined with those. However, it seems significantly better to do one of the hashing operations after this bitmixing rather than doing both hashes before the cheap bitmix.